lib/concurrent/synchronization/java_pure_object.rb

```diff
@@ -32,6 +33,10 @@ def ns_signal
           JRuby.reference0(self).notify
           self
         end
+
+        def ensure_ivar_visibility!
+          # relying on undocumented behavior of JRuby, ivar access is volatile
```

Comment:
> But saying we're always volatile now means we can't improve ivar performance by removing volatility later.

I'd say promising volatile for ivars would be a disastrous decision, that we are going to regret in the future.

PS: I'm not sure JRuby's i-vars are exactly volatile per JVM's spec. Yes, an i-var read always involves a volatile field read, but this is not true for writes, not every i-var write results in a proper Java volatile field write. In practice we don't run into trouble because of x86 and the current JRuby performance/overhead.